### DISCUSSION TITLE
When to use provider vs websocket provider?

### DISCUSSION THREAD

**Discussion by frolic** [wevm/wagmi#517] [Q&A]
I just realized I was eating up my Alchemy API limit due to polling from `provider.on('block', ...)` when I probably should have used a `websocketProvider.on('block', ...)`. I noticed that `useBlockNumber` already accounts for this and uses websocket provider when available.

Is there a distinction to be aware of when calling contracts, etc.? Should we ~always use a websocket provider if available?

Another thing I've noticed is that if I use a single websocket provider for a long-running thing, and the device goes offline, the websocket is closed and provider is stuck in a closed state. Is there a way to detect that? Requesting a new provider each time I need it seems ~wrong, since it'll establish a new WS connection.

**Comment by PDiTO** [comment]
Bumping this in case anyone has any information. Searched around for a day but found very little on best use cases here.
